extensions/amp-sticky-ad/0.1/amp-sticky-ad.js

```diff
@@ -28,15 +30,71 @@ class AmpStickyAd extends AMP.BaseElement {
   isLayoutSupported(layout) {
     return layout == Layout.NODISPLAY;
   }
-
-  /** @override */
+  /**
+   * @override
+   * @this {undefined}  // Make linter happy
+   */
   buildCallback() {
     /** @const @private {boolean} */
     this.isExperimentOn_ = isExperimentOn(this.getWin(), TAG);
     if (!this.isExperimentOn_) {
       dev.warn(TAG, `TAG ${TAG} disabled`);
       return;
     }
+    this.viewport_ = this.getViewport();
+    this.isDisplayed_ = false;
+    this.initialScrollTop_ = this.viewport_.getScrollTop();
+    this.scrollHeight_ = this.viewport_.getScrollHeight();
+    this.viewportHeight_ = this.viewport_.getSize().height;
+    /** @const @private {!Vsync} */
+    this.vsync_ = vsyncFor(this.getWin());
+    this.viewport_.onScroll(() => {
+      if (!this.isDisplayed_) {
+        const scrollDist =
+            (this.viewport_.getScrollTop() - this.initialScrollTop_);
+        if (this.viewportHeight_ < Math.abs(scrollDist)) {
+          // scroll down
+          if (scrollDist < 0) {
+            if (this.viewport_.getScrollTop() < this.viewportHeight_) {
+              // Not sure about this
+              this.initialScrollTop_ = this.viewport_.getScrollTop();
+              return;
+            }
+          } /*scroll up*/else {
+            const remainHeight = this.scrollHeight_
+                - this.viewport_.getScrollTop() - this.viewportHeight;
+            if (remainHeight < this.viewportHeight_) {
+              this.initialScrollTop_ = this.viewport_.getScrollTop();
+              return;
+            }
+          }
+          this.isDisplayed_ = true;
+          this.element.style.display = 'block';
+          this.vsync_.mutate(() => {
+            setStyles(this.element, {
+              'display': 'block',
+            });
+            this.viewport_.addToFixedLayer(this.element);
+            this.scheduleLayout(this.getRealChildren());
+          });
+        }
+      }
+    });
+  }
+
+  /**
+   * @override
+   * @this {undefined}  // Make linter happy
+   */
+  layoutCallback() {
+    this.isExperimentOn_ = isExperimentOn(this.getWin(), TAG);
+    if (!this.isExperimentOn_) {
+      dev.warn(TAG, `TAG ${TAG} disabled`);
+      return Promise.resolve();
+    }
+    this.scheduleLayout(this.element.firstElementChild);
```

Comment:
Also, a "real child" being first in DOM order is not necessarily guaranteed to indicate that it will be visible first. What are the rules/constraints here?